src/dependencies.rs

```diff
@@ -0,0 +1,305 @@
+    pub name: String,
+    pub version: VersionSpec,
+    pub line: String,
+}
+
+impl SimpleDependency {
+    pub fn from_frozen(frozen: &FrozenDependency) -> Self {
+        let name = &frozen.name;
+        let line = format!("{}=={}", name, frozen.version);
+        let version = LockedDependency::parse_simple_version(&line);
+        SimpleDependency {
+            name: name.to_string(),
+            version,
+            line,
+        }
+    }
+
+    pub fn bump(&mut self, new_version: &str) -> bool {
+        let VersionSpec { start, end, value } = &self.version;
+        if new_version == value {
+            return false;
+        }
+        self.line = format!(
+            "{}{}{}",
+            &self.line[0..*start],
+            new_version,
+            &self.line[*end..],
+        );
+        self.version.value = new_version.to_string();
+        true
+    }
+
+    pub fn freeze(&mut self, new_version: &str) {
+        self.bump(new_version);
+    }
+}
+
+#[derive(Debug)]
+pub enum LockedDependency {
+    Git(GitDependency),
+    Simple(SimpleDependency),
+}
+
+impl LockedDependency {
+    pub fn line(&self) -> String {
+        match self {
+            LockedDependency::Git(x) => x.line.to_string(),
+            LockedDependency::Simple(x) => x.line.to_string(),
+        }
+    }
+
+    pub fn name(&self) -> String {
+        match self {
+            LockedDependency::Git(x) => x.name.to_string(),
+            LockedDependency::Simple(x) => x.name.to_string(),
+        }
+    }
+
+    pub fn from_line(line: &str) -> Result<LockedDependency, ParseError> {
+        if line.contains("#egg=") {
+            let name = Self::parse_git_name(&line);
+            let git_ref = Self::parse_git_ref(&line)?;
+            let dep = GitDependency {
+                line: line.to_string(),
+                name,
+                git_ref,
+            };
+            return Ok(LockedDependency::Git(dep));
+        }
+        if line.contains("==") {
+            let version = Self::parse_simple_version(&line);
+            let name = Self::parse_simple_name(&line);
+            let dep = SimpleDependency {
+                line: line.to_string(),
+                name,
+                version,
+            };
+            return Ok(LockedDependency::Simple(dep));
+        }
+        Err(ParseError::new("neither a simple dep nor a git dep"))
+    }
+
+    fn parse_simple_name(line: &str) -> String {
+        let words: Vec<_> = line.split("==").collect();
+        let dep_name = words[0];
+        dep_name.trim().to_string()
+    }
+
+    fn parse_simple_version(line: &str) -> VersionSpec {
+        let equal_index = line
+            .find("==")
+            .unwrap_or_else(|| panic!("'{}' should contain '=='", line));
+        let mut start = equal_index + 2;
+        let mut end = line.len();
+        let colon_index = line.find(';');
+        if let Some(colon_index) = colon_index {
+            end = colon_index;
```

Comment:
Curious what happens if I give ";==" as input! 
I think instead of a parse error you panic with a 'begin > end' assert

Reply:
(That's when I'm glad this isn't C++!)

Reply:
The lines here come directly from `pip freeze` so ...

Plus, rust will give exactly this panic message if it happens :p

Reply:
Haha fair enough!